### PR TITLE
Support global-workflow using Rocky 8 on CSPs

### LINKED ISSUE #982: Add Jobs and Scripts for JEDI-Based Aerosol DA
**Description**
To support global aerosol initialization for future GEFS/GFS implementations, we need to add to global-workflow jobs and scripts similar to what have already been created for JEDI-based atmospheric variational DA, but for a separate, aerosol analysis.

**Requirements**
See #981 but I think we need 3 new jobs/scripts:
- aeroanalprep
- aeroanalrun
- aeroanalpost

The prep job should:
- convert netCDF VIIRS AOD obs to IODA format using python/compiled code
- stage backgrounds, fix files, and all other things necessary to run an aerosol analysis.

Run job:
- create YAML file from template(s)
- Run fv3jedi_var.x

Post job:
- archive diagnostic files
- move/copy increments/analysis files and make changes as needed to stage for model initialization

**Acceptance Criteria (Definition of Done)**
A 'dummy' cycling system is able to run with Aerosol DA tasks included.

**Dependencies**
This will depend on #981 before testing/debugging can properly occur

### LINKED ISSUE #2997: support global-workflow on CSPs with Rocky 8
### What new functionality do you need?

With Rocky 8 is default OS on CSPs, and ParallelWorks only support Rocky 8 starting 1/1/2025.
We need to move global-workflow to Rocky 8 and make sure it works before 1/1/2025.

### What are the requirements for the new functionality?

Need Rocky 8 supported spack-stack on CSPs, and EPIC's SI team has already got it available on CSPs.

### Acceptance Criteria

-[x] global-workflow compile successfully on CSPs, which includes all its components.
-[x] global-workflow test case run succeed on CSPs.

### Suggest a solution (optional)

Modify module files to use Rocky 8 supported spack-stack.

### PR DESCRIPTION
**Pull request by weihuang-jedi**
# Description

With ParallelWorks now default Rocky 8 on CSPs, and move to Rocky 8 only after 1/1/2025,
we need to modify global-workflow module files to use Rocky 8 supported spack-stack,
and test compile and run to make sure all works under Rocky 8.

i) Rocky 8 update new features:

a. Wave worked in C48_S2SWA_gefs case, so turn SUPPORT_WAVES to "YES" in awspw.yaml.
        Actually, if we did not set SUPPORT_WAVES to "YES", setup_expt.py will rise exception.

b. Using two type of nodes (chips/queues) on AWS, compute/process, where forecasts run in "compute" queue,
        which is a big node (more cores), others run in "process" queue, which has small node (less cores).

ii) Rocky 8 update needs the following submodules at or newer than the tags below.

1. gfs_utils:

commit 4848ecbb5e713b16127433e11f7d3edc6ac784c4 (HEAD, origin/develop, origin/HEAD, develop)
Author: Wei Huang <wei.huang@noaa.gov>
Date:   Fri Oct 18 10:41:25 2024 -0600

    Make gfs-utils compile on CSPs with Rocky 8 (#81)

    Support Rocky 8 on CSPs.

2. ufs_utils:

commit 23237610845c3a4438b21b25e9b3dc25c4c15b73 (HEAD)
Author: Wei Huang <wei.huang@noaa.gov>
Date:   Wed Oct 9 11:55:13 2024 -0600

    Support UFS_UTILS on CSPs under Rocky 8 (#989)

    Fixes #982.

3. upp:

commit 66a422db80ea129dd87285fe6e811d4b6e1fe29b (HEAD)
Author: Wei Huang <wei.huang@noaa.gov>
Date:   Wed Oct 2 14:38:22 2024 -0600

    Make UPP works with Rocky 8 on CSPs (#1034)

    * Make UPP works with Rocky 8 on CSPs

    * Remove unneeded path

    * simplify modulefile

4. ufs_model:

commit 29c2703c715ebdb47bbd4bcc811db340eae530e5 (HEAD)
Author: Cameron Book <43379611+ulmononian@users.noreply.github.com>
Date:   Tue Nov 12 13:08:12 2024 -0800

    Add developmental test cases: idealized baroclinic wave and 2020 July CAPE cases + https://github.com/ufs-community/ufs-weather-model/pull/2459 (#2461)

    * UFSWM - Add tests-dev ATM-only idealized dry baroclinic wave test and a 2020 July CAPE case
    * UFSWM - Update modulefile to support Rocky 8 on CSPs, with ParallelWorks

    ---------

    Co-authored-by: Wei Huang <wei.huang@noaa.gov>
    Co-authored-by: Jong Kim <jong.kim@noaa.gov>

  Resolves #2997

# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics

- Does this change require a documentation update? YES/NO

  - [x] GFS-utils #81
  - [x] UFS-utils UFS_UTILS#989 (already merged)
  - [x] UFS-weather-model #2456
  - [ ] UPP #1034  (already merged)

# How has this been tested?

- Clone and build on CSPs
- Forecast-only on AWS
- GEFS test on AWS

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] I have made corresponding changes to the system documentation if necessary

### CODE REVIEW

**Comment by weihuang-jedi** [comment]
Show tasks run on different partitions: compute and process

             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
                92   compute c48gefs_ Wei.Huan  R       9:04      1 weihuang-weisfsemcaws-00056-1-0002
                84   compute c48gefs_ Wei.Huan  R      10:03      1 weihuang-weisfsemcaws-00056-1-0003
                85   compute c48gefs_ Wei.Huan  R      10:03      1 weihuang-weisfsemcaws-00056-1-0004
                12   compute c48atm_g Wei.Huan  R      23:45      1 weihuang-weisfsemcaws-00056-1-0001
               164   process c48gefs_ Wei.Huan CF       0:03      1 weihuang-weisfsemcaws-00056-2-0019
               165   process c48gefs_ Wei.Huan CF       0:03      1 weihuang-weisfsemcaws-00056-2-0020
               148   process c48gefs_ Wei.Huan CF       2:04      1 weihuang-weisfsemcaws-00056-2-0018
               158   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0030
               159   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0001
               160   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0002
               161   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0003
               162   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0004
               163   process c48gefs_ Wei.Huan  R       0:03      1 weihuang-weisfsemcaws-00056-2-0005
               155   process c48atm_g Wei.Huan  R       0:05      1 weihuang-weisfsemcaws-00056-2-0017
               156   process c48atm_g Wei.Huan  R       0:05      1 weihuang-weisfsemcaws-00056-2-0022
               157   process c48gefs_ Wei.Huan  R       0:05      1 weihuang-weisfsemcaws-00056-2-0029

File: modulefiles/module_base.noaacloud.lua
Comment (WalterKolczynski-NOAA) [inline modulefiles/module_base.noaacloud.lua:8]: Is there bootstrapping or something going on here?

File: parm/config/gfs/config.base
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.base]: #3053 moved the machine-specific code out of `config.base`. Please do not add it back. Any machine-specific settings should go in the appropriate `workflow/hosts` file.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
This UFS model hash is before the one that introduced the issue with building wave pre/post executables (see #3110), so it should be safe to update.

File: parm/config/gfs/config.base
Comment (weihuang-jedi) [reply parm/config/gfs/config.base]: I'll remove it accordingly.

File: modulefiles/module_base.noaacloud.lua
Comment (weihuang-jedi) [reply modulefiles/module_base.noaacloud.lua:8]: @WalterKolczynski-NOAA Do you mean why we load/unload gnu module here?

Rahul ask the question on other PRs why we load/unload gnu, EPIC SI teams answer is that gnu is needed to load intel-stack, and intel-mpi-stack correctly. Since G-W does not use gnu, so gnu is unloaded after intel and mpi loaded correctly.

File: workflow/hosts/awspw.yaml
Comment (DavidHuber-NOAA) [inline workflow/hosts/awspw.yaml]: Since `WW3` will now be supported on all platforms, I think we can remove `SUPPORT_WAVES` from the host files and `setup_expt.py`.

```suggestion
```

File: workflow/hosts/azurepw.yaml
Comment (DavidHuber-NOAA) [inline workflow/hosts/azurepw.yaml]:
```suggestion
```

File: workflow/hosts/googlepw.yaml
Comment (DavidHuber-NOAA) [inline workflow/hosts/googlepw.yaml]:
```suggestion
```

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]

**Comment by DavidHuber-NOAA** [comment]
This block

https://github.com/NOAA-EMC/global-workflow/blob/ea07117c901c02b92e59d7c2eb3d3283fa953db0/workflow/setup_expt.py#L374-L385

Can be simplified to

```python
 machine = host.machine 
 for attr in ['resdetatmos', 'resensatmos']: 
     try: 
         expt_res = f'C{getattr(inputs, attr)}' 
     except AttributeError: 
         continue 
     if expt_res not in supp_res: 
         raise NotImplementedError(f"Supported resolutions on {machine} are:\n{', '.join(supp_res)}") 
```

File: modulefiles/module_base.noaacloud.lua
Comment (WalterKolczynski-NOAA) [reply modulefiles/module_base.noaacloud.lua:8]: That's not the case on any of the on-prem machines, which is why I asked.

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

File: env/AWSPW.env
Comment (DavidHuber-NOAA) [inline env/AWSPW.env]:
```suggestion
    export sys_tp="AWSPW"
```

File: env/AWSPW.env
Comment (DavidHuber-NOAA) [inline env/AWSPW.env]:
```suggestion
    export USE_CFP="YES"  # Use MPMD for downstream product generation on AWS
```

File: modulefiles/module_base.noaacloud.lua
Comment (DavidHuber-NOAA) [reply modulefiles/module_base.noaacloud.lua:8]: I'm guessing this has to do with the way Intel was built on the cloud.  The spack recipe requires gnu in order to get a valid glibc library to build against.  But the spack-written module files don't usually include it because they invariably use the system glibc library.  *shrug*

File: sorc/build_ufs.sh
Comment (DavidHuber-NOAA) [inline sorc/build_ufs.sh:15]: Thanks for this fix.

File: workflow/hosts/awspw.yaml
Comment (DavidHuber-NOAA) [inline workflow/hosts/awspw.yaml]: Is 192 not supported yet?

File: env/AWSPW.env
Comment (weihuang-jedi) [reply env/AWSPW.env]: Changed to AWS. good catch.

File: workflow/hosts/awspw.yaml
Comment (weihuang-jedi) [reply workflow/hosts/awspw.yaml]: I won't say that, but I have not run any C192 case yet. 
Let me add it there.

File: env/AWSPW.env
Comment (weihuang-jedi) [reply env/AWSPW.env]: corrected.

File: parm/config/gefs/config.resources
Comment (DavidHuber-NOAA) [inline parm/config/gefs/config.resources:52]: Looking at the table at the top of [Configuring NOAA Cloud Service Providers](https://global-workflow.readthedocs.io/en/latest/noaa_csp.html#configuring-noaa-cloud-service-providers), it looks like these could be
```bash
    export PARTITION_BATCH="compute"
    max_tasks_per_node=72
    ;;
  "AZUREPW")
    export PARTITION_BATCH="compute"
    max_tasks_per_node=48
    ;;
  "GOOGLEPW")
    export PARTITION_BATCH="compute"
    max_tasks_per_node=60
```

That said, there's no harm (except possibly cost) for running fewer tasks/node.  Just want to check.

File: workflow/hosts/azurepw.yaml
Comment (DavidHuber-NOAA) [inline workflow/hosts/azurepw.yaml:30]: Should 192 also be added for Azure and GCP?

**Review by DavidHuber-NOAA — DISMISSED** [review]
Changes look good.  Just a couple of questions.

File: parm/config/gefs/config.resources
Comment (DavidHuber-NOAA) [reply parm/config/gefs/config.resources:52]: If these are to be updated, then you will likely want to change parm/config/gfs/config.resources as well.

**Comment by aerorahul** [comment]
Once the questions are addressed and reviews obtained from @DavidHuber-NOAA @WalterKolczynski-NOAA, we need to run this PR on Hera, Hercules, and WCOSS2 as this PR updates the submodules for UFS weather model, ufs-utils and gfs-utils.

The AWSPW.env file also needs to remove the DA sections as that is not yet functional on the CSPs.

@weihuang-jedi When you have addressed the comments, please let @DavidHuber-NOAA @WalterKolczynski-NOAA know so they can review and run the CI on the on-prem HPCs

File: parm/config/gefs/config.resources
Comment (weihuang-jedi) [reply parm/config/gefs/config.resources:52]: The numbers I put there are physical Cores.
If we double the numbers, which means we are using vCPUs, I have not tested yet.
Let me run a test and see how that goes.

Comment (DavidHuber-NOAA) [reply parm/config/gefs/config.resources:52]: Ah, I see.  No need to run that test.  The global workflow does not support hyperthreading.

Comment (weihuang-jedi) [reply parm/config/gefs/config.resources:52]: Tried c7i.24xlarge, which AWS/PW said it has 96 vCPUs, if I set 
    npe_node_max=96
    max_tasks_per_node=96
    
    Then we get errors:
    
sbatch: error: Batch job submission failed: Requested node configuration is not available
2024-12-04 20:52:02 +0000 :: mgmt-weihuang-awsepicweirocky8c7i24xlarge-00013 :: Submitting gfs_fcst_seg0
2024-12-04 20:52:02 +0000 :: mgmt-weihuang-awsepicweirocky8c7i24xlarge-00013 :: Submission status of gfs_fcst_seg0 is pending at druby://10.1.0.105:45141
2024-12-04 20:53:02 +0000 :: mgmt-weihuang-awsepicweirocky8c7i24xlarge-00013 :: Submission status of previously pending gfs_fcst_seg0 is failure!  sbatch: error: CPU count per node can not be satisfied

Comment (DavidHuber-NOAA) [reply parm/config/gefs/config.resources:52]: Noted, thanks.  This should stay as is, then.

**Review by DavidHuber-NOAA — DISMISSED** [review]
Approved pending successful CI testing.

**Comment by WalterKolczynski-NOAA** [comment]
Failing due to #3121 There is a solution there that can be incorporated into this PR to prevent it from being blocked until #3032 is merged.

File: parm/config/gefs/config.resources
Comment (weihuang-jedi) [reply parm/config/gefs/config.resources:52]: Those are virtual CPU numbers. GW wants the real Core number, if uses virtual core, GW won't run.

**Comment by weihuang-jedi** [comment]
I have run the code (after sync with EMC develop) on AWS, both c48atm, and c48gefs run successfully.

**Comment by aerorahul** [comment]
> Experiment C96C48_ufs_hybatmDA **FAILED** on Hera in Build# 2 with error logs:
> 
> ```
> /scratch1/NCEPDEV/global/CI/2998/RUNTESTS/COMROOT/C96C48_ufs_hybatmDA_9e666f09/logs/2024022400/enkfgdas_atmensanlobs.log
> ```
> 
> Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/88b25d5f440f316dc1689b54b48a9b87)

@RussTreadon-NOAA @DavidNew-NOAA 
Hercules CI does not run this test.
Can we get a quick look at this failed log file (see link above)?

**Comment by DavidNew-NOAA** [comment]
This PR is changing the results of the first half-cycle ensemble forecast so that the input background states for the LETFK observer are diverging sufficiently from the test reference to trigger a failure. You can compare the test output and test reference here:

output: ```/scratch1/NCEPDEV/global/CI/2998/global-workflow/sorc/gdas.cd/build/gdas/test/testoutput/C96C48_ufs_hybatmDA_lgetkf_observer.test.out```
reference: ```/scratch1/NCEPDEV/global/CI/2998/global-workflow/sorc/gdas.cd/test/testreference/C96C48_ufs_hybatmDA_lgetkf_observer.ref```

You can just diff the two files to see the difference. The 3dvar test reference also has differences. I'm not sure if they would trigger a test failure. Perhaps the lgetkf observer job finished first before the 3dvar job and hence triggered a CI failure first.

This will require a quick GDASApp PR to update the reference files. I don't understand why the outputs of the forecast are changing though. I don't see that the UFS hash is being updated in this PR.

**Comment by DavidNew-NOAA** [comment]
These are the lines in the log file that indicate a regression test failure: 
```
 0: terminate called after throwing an instance of 'oops::TestReferenceFloatMismatchError'
 0:   what():  Test reference Float mismatch @ Line:10
 0: Test Val : 2.2505994886159897e-02
 0: Ref  Val : 2.2395884618163109e-02
 0: Delta    : 1.1011026799678802e-04
 0: Relative tolerance: 2.2450939752161502e-05
 0: Absolute tolerance: 1.0000000000000001e-05
 0: Test Line: 'water_vapor_mixing_ratio_wrt_moist_air       | Min:+1.6836553484722572e-08 Max:+2.2505994886159897e-02 RMS:+5.0110427820345893e-03'
 0: Ref Line : 'water_vapor_mixing_ratio_wrt_moist_air       | Min:+1.6847316430812498e-08 Max:+2.2395884618163109e-02 RMS:+5.0107896876815721e-03'
```

**Comment by RussTreadon-NOAA** [comment]
Thank you @DavidNew-NOAA for looking at the output.  As David notes, the reference check failed not due to DA but rather due to changes in the background used by `H(x)` in enkfgdas_atmensanlobs.

Do we expect this PR to alter forecast fields (deterministic and/or ensemble)?  The modeling team should confirm that observed differences in model output are acceptable.

**Comment by RussTreadon-NOAA** [comment]
This PR does not change the `sorc/ufs_model.fd` hash.  g-w PR #3145 updated the `sorc/ufs_model.fd` hash.  This updated hash is included in this PR.  

It was [noted](https://github.com/NOAA-EMC/global-workflow/pull/3163#issuecomment-2543171278) in g-w PR #3163 that the updated `sorc/ufs_model.fd` hash altered forecast output.  GDASApp reference files have been updated and are included in the updated `sorc/gdas.cd` hash in g-w PR #3163.   As documented in g-w PR #3163, g-w CI passes on WCOSS2 (Dogwood), Hera, and Orion.

One path forward is to 
1. merge g-w PR [#3163](https://github.com/NOAA-EMC/global-workflow/pull/3163) into g-w `develop`
2. update `NOAA-EPIC:csps-rocky8` with the updated g-w `develop`
3. rerun Hera CI for this PR

**Review by aerorahul — APPROVED** [review]
looks good.  tests have passed on Hera and Hercules.

**Comment by WalterKolczynski-NOAA** [comment]
gdas build fails on WCOSS, but it fails in develop too

**Comment by WalterKolczynski-NOAA** [comment]
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_2998/RUNTESTS on WCOSS

**Pull request merged**
